private/pkg/manifest/manifest.go

```diff
@@ -0,0 +1,269 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Manifests are a list of paths and their hash digests, canonically ordered by
+// path in increasing lexographical order. Manifests are encoded as:
+//
+//	<hash type>:<digest>[SP][SP]<path>[LF]
+//
+// "shake256" is the only supported hash type. The digest is 64 bytes of hex
+// encoded output of SHAKE256. See golang.org/x/crypto/sha3 and FIPS 202 for
+// details on the SHAKE hash.
+package manifest
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"go.uber.org/multierr"
+	"golang.org/x/crypto/sha3"
+)
+
+var errPartial error = errors.New("partial record")
+
+const shake256Name = "shake256"
+
+type Digest struct {
+	dtype  string
+	digest []byte
+	hexstr string
+}
+
+func NewDigestFromBytes(dtype string, digest []byte) *Digest {
+	return &Digest{
+		dtype:  dtype,
+		digest: digest,
+		hexstr: hex.EncodeToString(digest),
+	}
+}
+
+func NewDigestFromHex(dtype string, hexstr string) (*Digest, error) {
+	digest, err := hex.DecodeString(hexstr)
+	if err != nil {
+		return nil, err
+	}
+	return NewDigestFromBytes(dtype, digest), nil
+}
+
+func NewDigestFromString(typedDigest string) (*Digest, error) {
+	hashfunc, digestStr, found := strings.Cut(typedDigest, ":")
+	if !found {
+		return nil, errors.New("malformed digest string")
+	}
+	return NewDigestFromHex(hashfunc, digestStr)
+}
+
+func (d *Digest) String() string {
+	return d.Type() + ":" + d.Hex()
+}
+
+func (d *Digest) Type() string {
+	return d.dtype
+}
+
+func (d *Digest) Bytes() []byte {
+	return d.digest
+}
+
+func (d *Digest) Hex() string {
+	return d.hexstr
+}
+
+// ManifestError occurs when a manifest is malformed.
+type ManifestError struct {
+	lineno int
+	msg    string
+}
+
+var _ error = (*ManifestError)(nil)
+
+func newManifestError(lineno int, msg string) *ManifestError {
+	return &ManifestError{
+		lineno: lineno,
+		msg:    msg,
+	}
+}
+
+func (e *ManifestError) Error() string {
+	return fmt.Sprintf("invalid manifest: %d: %s", e.lineno, e.msg)
+}
+
+// Manifest represents a list of paths and their digests.
+type Manifest struct {
+	paths   map[string]*Digest
+	digests map[string]string
+	hash    sha3.ShakeHash
+}
+
+var _ encoding.TextMarshaler = (*Manifest)(nil)
+var _ encoding.TextUnmarshaler = (*Manifest)(nil)
+
+// NewManifest creates an empty manifest.
+func NewManifest() *Manifest {
+	return &Manifest{
+		paths:   make(map[string]*Digest),
+		digests: make(map[string]string),
```

Comment:
nit: Maybe name `pathToDigest` and `digestToPath`?

Reply:
Or even the switching its names, `paths` map doesn't hold paths, holds digests, accessed by it's path haha.

Reply:
It's all about perspective! I'll make the names more approachable. :)